ls1intum/Artemis
# Bugfix/Improve result and submission saving for programming exercises + write integration tests

### Checklist

### Motivation and Context
Some results for programming exercises do not appear in Artemis. We assume the problem is that they are not stored correctly in the database. There might be an issue with the one to one association between submission and result and the orphan removal and cascade annotations.

### Description
In this pull request, I removed the orphan removal and cascade annotations and explicitly set the bidirectional association between submission and result when Bamboo notifies Artemis about new results.

_Defined scenarios of the programming submissions:_

**1) The student makes a submission from the code-editor or directly from the git repository (default scenario):**
- VCS notifies the Artemis server about new commit: A programming submission with the commitHash of the student commit and type `MANUAL` is created.
- CI notifies the Artemis server about new build result: A result is created and assigned to the submission.

**2) Commit to the Test Repository by the instructor:**
- VCS notifies the Artemis server about new commit: For every participation (including template & solution!) a programming submission is created with the commitHash of the test repository commit and the type `TEST`.
- CI notifies the Artemis server about new build result: A result is created and assigned to the submission.

**3) Manual build trigger from the CI interface:**
- VCS notifies the Artemis server about new commit: This does not happen, as no commit is made!
- CI notifies the Artemis server about new build result: A submission with the commitHash from the build result and type `OTHER` and a result are created and linked together.

Please note: Scenario 3 will also be used when no submission was created out of an unknown error! I think it is most important that we try to persist every result that comes in.

**Issues found when refactoring:**
- [x] A new commit in the test repository causes submissions to be created for every participation in the ProgrammingSubmissionResource. In this case the commitHash of the Test Repository is used as an identifier for the submissions.
However the ResultResource always uses the commitHash of the Assignment Repository and therefore doesn't find the created submissions from the test submission. => I changed it so that we try to find out what kind of result came in (from a test submission, a manual build, etc.) based on the last pending submission.
- [x] Since the Remodel Participations PR, there were no longer submissions created for template and solution participations on a test submission. => I added them to the array that is iterated.
- [x] There were scenarios in the ResultResource where a second result was created for the same submission. This causes an IntegrityViolation. However as the ResultResource hat the `@Transactional` annotation, a success message with the result - that could never be saved - was still send to the user. => I removed the annotation, refactored the Resource & Service and now check if there is already a result to avoid the error. 
- [x] Fix issue with failing test that runs alone without issues.

**The issues might be too big for this PR:**
- [ ] Can we improve the build feedback of Bamboo so that we know _how_ a build was triggered? (Manual, Test, Normal Commit)
- [ ] Can we remove the automatic build trigger from test to other repos? I can image that this can cause issues as Artemis might be low on resources and can't answer requests to create submissions, but the results are already coming in.


### Steps for Testing
1. Log in to ArTEMiS
2. Navigate to Course Administration
3. Create a programming exercise
4. Start the exercise, clone the repo, commit and push code
5. Check if the result is inserted correctly

We also need to check submission and result creation for all other exercises, because the association between submission and result has changed

### TODOs

**Additional test cases:**
1. [x] Mock that Bitbucket notifes Artemis about a submission for a participation that does not exist and verify that no submission is created.
2. [x] Mock that Bitbucket notifies Artemis about a new submission and check that it is stored correctly
3. [x] New Submission (see 1) + Mock that Bamboo notifies Artemis about a new result due to one new student commit and check that it is stored correctly and check that it is stored correctly
4. [x] New Submission (see 1) + Mock that Bamboo notifies Artemis about a new result due to two new student commits and check that it is stored correctly
5. [x] New Submission (see 1) + Mock that Bamboo notifies Artemis about a new result due to a commit in the test repo and check that it is stored correctly
6. [x] Mock that Bamboo notifies Artemis about a new result due to a manual build and check that it is stored correctly
7. [x] New Submission (see 1) + Mock that Bamboo notifies Artemis about a new result in a template participation and check that it is stored correctly
8. [x] New Submission (see 1) + Mock that Bamboo notifies Artemis about a new result in a solution participation and check that it is stored correctly
9. [x] The same submission with the same commit hash should not create two submissions.
